plugins/newsql/newsql_evbuffer.c

```diff
@@ -153,7 +153,7 @@ static int newsql_done_cb(struct sqlclntstate *clnt)
         appdata->query = NULL;
     }
     if (sql_done(appdata->writer) == 0) {
-        event_once(appsock_timer_base, newsql_read_again, appdata);
+        newsql_read_again(-1, 0, appdata);
```

Comment:
Should this (disable heartbeat/timeout) be push down into `sql_done` which will performs it under lock so it doesn't race with heartbeat?

Reply:
`event_del` would block if the callback is being executed in the event base. Therefore it's safe to call `newsql_read_again()` here without holding `wr_lock` or scheduling the function via `appsock_timer_base`.

Reply:
👍